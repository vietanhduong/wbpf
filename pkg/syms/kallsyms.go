package syms

import (
	"fmt"
	"os"
	"runtime"
	"strings"
)

func parseKallsyms(path string) ([]Symbol, error) {
	f, err := os.OpenFile(path, os.O_RDONLY, 0)
	if err != nil {
		return nil, fmt.Errorf("os read file %s: %w", path, err)
	}
	defer f.Close()
	var ret []Symbol
	var kernelAddr uint64 = 0
	if runtime.GOARCH == "amd64" {
		// https://www.kernel.org/doc/Documentation/x86/x86_64/mm.txt
		kernelAddr = 0x00ffffffffffffff
	}
	for {
		var sym Symbol
		var perm, module string
		n, _ := fmt.Fscanf(f, "%x %1s %s %s\n", &sym.Start, &perm, &sym.Name, &module)
		if n != 4 && n != 3 {
			break
		}

		if sym.Start == 0 || sym.Start < kernelAddr {
			continue
		}

		if perm == "b" || perm == "B" ||
			perm == "d" || perm == "D" ||
			perm == "r" || perm == "R" {
			continue
		}

		sym.Module = "kernel"

		if module = strings.TrimSpace(module); module != "" &&
			module[0] == '[' &&
			module[len(module)-1] == ']' {
			sym.Module = module[1 : len(module)-1]
		}
		ret = append(ret, sym)
	}
	return ret, nil
}
