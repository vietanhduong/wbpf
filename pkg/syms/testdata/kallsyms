0000000000000000 A fixed_percpu_data
0000000000000000 A __per_cpu_start
0000000000001000 A cpu_debug_store
0000000000002000 A irq_stack_backing_store
0000000000006000 A cpu_tss_rw
000000000000b000 A gdt_page
000000000002f958 A centrino_cpu
000000000002f960 A centrino_model
000000000002f980 A cpuidle_dev
000000000002fca0 A cpuidle_devices
000000000002fcc0 A ladder_devices
000000000002fe00 A menu_devices
000000000002fe80 A teo_cpus
000000000002ff40 A cpu_trig
000000000002ff80 A numachip2_ced
0000000000030080 A stimer0_evt
0000000000030088 A dtpm_per_cpu
0000000000030090 A idle_inject_device
00000000000300a0 A idle_inject_thread
00000000000300c0 A napi_alloc_cache
00000000000302e0 A netdev_alloc_cache
0000000000030300 A __net_cookie
0000000000030320 A flush_works
0000000000030340 A bpf_sp
0000000000030540 A bpf_redirect_info
0000000000030580 A __sock_cookie
00000000000305c0 A netpoll_srcu_srcu_data
0000000000030740 A dm_hw_cpu_data
00000000000307c0 A dm_cpu_data
0000000000030840 A sch_frag_data_storage
0000000000030890 A nf_skb_duplicated
0000000000034000 A __per_cpu_end
ffffffffb5000000 T startup_64
ffffffffb5000000 T _stext
ffffffffb5000000 T _text
ffffffffb5000070 T secondary_startup_64
ffffffffb5000075 T secondary_startup_64_no_verify
ffffffffb5000140 t verify_cpu
ffffffffb5000240 T sev_verify_cbit
ffffffffb50002a0 T start_cpu0
ffffffffb50002ac T vc_boot_ghcb
ffffffffb5000320 T __startup_64
ffffffffb5003310 t perf_trace_initcall_finish
ffffffffc035f1d0 t autofs_dev_ioctl [autofs4]
ffffffffc035f1f0 t autofs_dev_ioctl_compat [autofs4]
ffffffffc035f210 t find_autofs_mount.isra.0 [autofs4]
ffffffffc035f2e0 t autofs_dev_ioctl_ismountpoint [autofs4]
ffffffffc035f470 t autofs_dev_ioctl_requester [autofs4]
ffffffffc035f580 t autofs_dev_ioctl_openmount [autofs4]
ffffffffc0364160 d _autofs_dev_ioctl_misc [autofs4]
ffffffffc0360c50 r __func__.34560 [autofs4]
ffffffffc0360c80 r _dev_ioctl_fops [autofs4]
ffffffffc03648c0 d __this_module [autofs4]
ffffffffc035f6ab t cleanup_module [autofs4]
ffffffffc035b2d0 t autofs_free_ino [autofs4]
ffffffffc035b2a0 t autofs_clean_ino [autofs4]
ffffffffc035b300 t autofs_kill_sb [autofs4]
ffffffffc0360600 r autofs_dentry_operations [autofs4]
ffffffffc035e6d0 t autofs_expire_wait [autofs4]
ffffffffc035d2f0 t autofs_wait_release [autofs4]
ffffffffc035e990 t autofs_do_expire_multi [autofs4]
ffffffffc035d0e0 t is_autofs_dentry [autofs4]
ffffffffc0360860 r autofs_root_operations [autofs4]
ffffffffc035d1f0 t autofs_catatonic_mode [autofs4]
ffffffffc035b240 t autofs_new_ino [autofs4]
ffffffffc035d6b0 t autofs_wait [autofs4]
ffffffffc035b380 t autofs_get_inode [autofs4]
ffffffffc0360740 r autofs_dir_operations [autofs4]
ffffffffc0360680 r autofs_dir_inode_operations [autofs4]
ffffffffc035b470 t autofs_fill_super [autofs4]
ffffffffc035eb60 t autofs_expire_multi [autofs4]
ffffffffc0364040 d autofs_fs_type [autofs4]
ffffffffc035e820 t autofs_expire_run [autofs4]
ffffffffc0360980 r autofs_symlink_inode_operations [autofs4]
ffffffffc035f690 t autofs_dev_ioctl_exit [autofs4]
ffffffffc037ee4c t bpf_prog_6deef7357e7b4530 [bpf]
