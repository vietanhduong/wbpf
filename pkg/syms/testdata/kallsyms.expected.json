[
  { "start": 18446744072451260416, "name": "startup_64", "module": "kernel" },
  { "start": 18446744072451260416, "name": "_stext", "module": "kernel" },
  { "start": 18446744072451260416, "name": "_text", "module": "kernel" },
  { "start": 18446744072451260528, "name": "secondary_startup_64", "module": "kernel" },
  { "start": 18446744072451260533, "name": "secondary_startup_64_no_verify", "module": "kernel" },
  { "start": 18446744072451260736, "name": "verify_cpu", "module": "kernel" },
  { "start": 18446744072451260992, "name": "sev_verify_cbit", "module": "kernel" },
  { "start": 18446744072451261088, "name": "start_cpu0", "module": "kernel" },
  { "start": 18446744072451261100, "name": "vc_boot_ghcb", "module": "kernel" },
  { "start": 18446744072451261216, "name": "__startup_64", "module": "kernel" },
  { "start": 18446744072451273488, "name": "perf_trace_initcall_finish", "module": "kernel" },
  { "start": 18446744072639345104, "name": "autofs_dev_ioctl", "module": "autofs4" },
  { "start": 18446744072639345136, "name": "autofs_dev_ioctl_compat", "module": "autofs4" },
  { "start": 18446744072639345168, "name": "find_autofs_mount.isra.0", "module": "autofs4" },
  { "start": 18446744072639345376, "name": "autofs_dev_ioctl_ismountpoint", "module": "autofs4" },
  { "start": 18446744072639345776, "name": "autofs_dev_ioctl_requester", "module": "autofs4" },
  { "start": 18446744072639346048, "name": "autofs_dev_ioctl_openmount", "module": "autofs4" },
  { "start": 18446744072639346347, "name": "cleanup_module", "module": "autofs4" },
  { "start": 18446744072639328976, "name": "autofs_free_ino", "module": "autofs4" },
  { "start": 18446744072639328928, "name": "autofs_clean_ino", "module": "autofs4" },
  { "start": 18446744072639329024, "name": "autofs_kill_sb", "module": "autofs4" },
  { "start": 18446744072639342288, "name": "autofs_expire_wait", "module": "autofs4" },
  { "start": 18446744072639337200, "name": "autofs_wait_release", "module": "autofs4" },
  { "start": 18446744072639342992, "name": "autofs_do_expire_multi", "module": "autofs4" },
  { "start": 18446744072639336672, "name": "is_autofs_dentry", "module": "autofs4" },
  { "start": 18446744072639336944, "name": "autofs_catatonic_mode", "module": "autofs4" },
  { "start": 18446744072639328832, "name": "autofs_new_ino", "module": "autofs4" },
  { "start": 18446744072639338160, "name": "autofs_wait", "module": "autofs4" },
  { "start": 18446744072639329152, "name": "autofs_get_inode", "module": "autofs4" },
  { "start": 18446744072639329392, "name": "autofs_fill_super", "module": "autofs4" },
  { "start": 18446744072639343456, "name": "autofs_expire_multi", "module": "autofs4" },
  { "start": 18446744072639342624, "name": "autofs_expire_run", "module": "autofs4" },
  { "start": 18446744072639346320, "name": "autofs_dev_ioctl_exit", "module": "autofs4" },
  { "start": 18446744072639475276, "name": "bpf_prog_6deef7357e7b4530", "module": "bpf" }
]
